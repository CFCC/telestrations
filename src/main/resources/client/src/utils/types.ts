export interface Page {
  type: "image" | "text";
  authorId: string;
  content: string;
  id: string;
}

export interface Notebook {
  originalOwnerId: string;
  pages: Page[];
}

export interface Settings {
  id: string;
  avatar: string;
  name: string;
  connected: boolean;
}

export interface Player {
  id: string;
  setting: Settings;
  notebookQueue: Notebook[];
  startOfTimeOffline: Date;
}

export interface Game {
  code: string;
  adminId: string;
  players: Player[];
  active: boolean;
  isDone: boolean;
}

export enum GameState {
  // Common
  LOGIN = "login",
  GAME_CODE = "game code",
  WAITING_TO_START = "waiting to start",

  // Server
  BIRDS_EYE = "bird's eye",
  SINGLE_PLAYER = "single player",
  PLAYER_HISTORY = "player history",
  NOTEPAD_HISTORY = "notepad history",

  // Client
  IN_GAME = "in game",
  FINISHED = "finished",
  WAITING_FOR_CONTENT = "waiting for content",
}
