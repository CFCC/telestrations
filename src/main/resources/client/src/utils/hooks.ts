import { FormEvent, useState } from "react";

export function useBoolean(
  initialState: boolean
): [boolean, () => void, () => void] {
  const [bool, setBool] = useState(initialState);
  const setBoolTrue = () => setBool(true);
  const setBoolFalse = () => setBool(false);

  return [bool, setBoolTrue, setBoolFalse];
}

export function useEvent<T>(
  defaultValue: T,
  getter: (...args: any) => T
): [T, (...args: any) => void, (newState: T) => void] {
  const [state, setState] = useState(defaultValue);
  const wrappedSetState = (...args: any) => setState(getter.apply(null, args));

  return [state, wrappedSetState, setState];
}

export function useInput<T extends string>(
  defaultValue: T
): [T, (e: FormEvent<HTMLInputElement | HTMLTextAreaElement> | T) => void] {
  const [state, setState] = useState(defaultValue);

  function onChange(e: FormEvent<HTMLInputElement | HTMLTextAreaElement> | T) {
    if (typeof e === "object") {
      setState(e.currentTarget.value as T);
    } else {
      setState(e);
    }
  }

  return [state, onChange];
}
