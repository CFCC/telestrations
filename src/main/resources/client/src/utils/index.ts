export function sleep(t: number): Promise<void> {
  return new Promise((res) => setTimeout(res, t));
}
