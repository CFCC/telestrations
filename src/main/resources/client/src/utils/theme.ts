import {createMuiTheme} from "@material-ui/core/styles";
import {createGlobalStyle} from "styled-components";

export const primary = "hsl(50, 100%, 65%)";
export const darkPrimary = "hsl(40, 100%, 45%)";
export const secondary = "hsl(210, 100%, 33%)";

function headers(styles: object): object {
    return {
        h1: styles,
        h2: styles,
        h3: styles,
        h4: styles,
        h5: styles,
        h6: styles,
    };
}

function body(styles: object): object {
    return {
        subtitle1: styles,
        subtitle2: styles,
        body1: styles,
        body2: styles,
        button: styles,
        caption: styles,
        overline: styles,
    }
}

export const theme = createMuiTheme({
    typography: {
        ...body({fontFamily: "Fira Sans"}),
        ...headers({fontFamily: "Mali"}),
    },
});

export const GlobalStyles = createGlobalStyle`
    body {
        margin: 0;
        overflow: hidden;
    }
`;
