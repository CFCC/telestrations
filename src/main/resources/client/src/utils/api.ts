import SockJS from "sockjs-client";
import { Client } from "@stomp/stompjs";
import _ from "lodash";
import { v4 as uuid } from "uuid";

import { store, actions, rejoinGame } from "./store";
import { Game, Settings } from "./types";

const client = new Client({
  webSocketFactory: () => new SockJS(`${location.origin}/websocket-server`),
});
client.onConnect = () => {
  client.subscribe("/topic/rejoin-game", ({ body }) => {
    store.dispatch(rejoinGame(body));
  });

  client.subscribe("/topic/games", ({ body }) => {
    const response = JSON.parse(body) as string[];
    store.dispatch(actions.handleGamesListMessage(response));
  });

  client.subscribe(`/user/topic/errors/client`, ({ body }) => {
    store.dispatch(actions.handleRequestException(body));
  });

  client.subscribe(`/user/topic/errors/server`, ({ body }) => {
    store.dispatch(actions.handleServerException(body));
  });

  client.subscribe(`/user/topic/successes`, ({ body }) => {
    store.dispatch(actions.handleSuccess(body));
  });
};

export const getGameCodes = (callback: (ids: string[]) => void) => {
  return firebase
    .firestore()
    .collection("games")
    .where("status", "==", "lobby")
    .onSnapshot(async function (snapshot) {
      const newGames = snapshot.docs.map((doc) => doc.id);
      callback(newGames);
    });
};
export const setGameCode = (gameCode: string, isClient: boolean = false) => {
  const game = firebase
    .firestore()
    .doc(`games/${gameCode}`) as DocumentReference<Partial<Game>>;

  game.onSnapshot(async (snapshot) => {
    const newGame = snapshot.data();
    if (!newGame) return;

    if (isClient) {
      if (newGame.status === "in progress") {
        store.dispatch(setGameState(GameState.IN_GAME));
      } else if (newGame.status === "finished") {
        store.dispatch(setGameState(GameState.FINISHED));
        const currentUser = firebase.auth().currentUser;
        if (currentUser?.isAnonymous) {
          await currentUser.delete();
        }
      }
    }

    store.dispatch(updateGame({ ...newGame, id: snapshot.id }));
  });
  game.collection("notepads").onSnapshot((snapshot) => {
    const notepads = _.mapValues(
      _.keyBy(snapshot.docs, "id"),
      (d) => d.data() as Notepad
    );
    store.dispatch(updateNotepads(notepads));
  });
  game.collection("players").onSnapshot(async (snapshot) => {
    const players = _.mapValues(
      _.keyBy(snapshot.docs, "id"),
      (d) => d.data() as Player
    );
    store.dispatch(updatePlayers(players));

    if (!isClient) {
      const {
        notepads,
        game: { status },
      } = store.getState().firebase;
      const gameIsOver =
        status === "in progress" &&
        Object.entries(players).every(
          ([pid, p]) => notepads[p.currentNotepad]?.ownerId === pid
        ) &&
        Object.values(notepads).every((n) => n.pages.length > 1);

      if (gameIsOver) await game.set({ status: "finished" }, { merge: true });
    } else {
      const { uid } = store.getState().client.user;
      const { currentNotepad, queue } = players[uid];

      if (!currentNotepad && queue?.length > 0) {
        const newQueue = [...queue];
        const newNotepad = newQueue.shift();

        await (firebase
          .firestore()
          .doc(`games/${gameCode}/players/${uid}`) as DocumentReference<
          Partial<Player>
        >).set(
          { queue: newQueue, currentNotepad: newNotepad },
          { merge: true }
        );
        store.dispatch(clientSlice.actions.setGameState(GameState.IN_GAME));
      }
    }
  });
};
export const startGame = async () => {
  const {
    firebase: {
      game: { id: gameCode },
      players,
    },
  } = store.getState();
  const playerIds = Object.keys(players);

  await Promise.all(
    playerIds.map(async (playerId, i) => {
      const notepadId = uuid();
      await (firebase
        .firestore()
        .doc(
          `games/${gameCode}/notepads/${notepadId}`
        ) as DocumentReference<Notepad>).set({ ownerId: playerId, pages: [] });
      await (firebase
        .firestore()
        .doc(`games/${gameCode}/players/${playerId}`) as DocumentReference<
        Partial<Player>
      >).set(
        {
          currentNotepad: notepadId,
          nextPlayer: playerIds[(i + 1) % playerIds.length],
          queue: [],
        },
        { merge: true }
      );
    })
  );
  await (firebase.firestore().doc(`games/${gameCode}`) as DocumentReference<
    Partial<Game>
  >).set({ status: "in progress" }, { merge: true });
};
export const setGuess = _.throttle(async (guess: string) => {
  const {
    client: { user },
    firebase: {
      game: { id: gameCode },
      players,
      notepads,
    },
  } = store.getState();
  const currentNotepadId = players[user.uid].currentNotepad;
  const currentNotepad = notepads[currentNotepadId];
  if (!currentNotepad) return;

  const pages = _.cloneDeep(currentNotepad.pages);
  const needsNewPage = _.last(pages)?.author !== user.uid;
  if (needsNewPage) {
    pages.push({
      author: user.uid,
      lastUpdated: new Date().getTime(),
      content: "",
    });
  }

  if (pages.length % 2 === 1) {
    pages[pages.length - 1].content = guess;
    pages[pages.length - 1].lastUpdated = new Date().getTime();
  } else {
    let fileName = _.last(pages)?.content;
    if (!fileName) {
      fileName = `${uuid()}.png`;
      pages[pages.length - 1].content = fileName;
    }

    pages[pages.length - 1].lastUpdated = new Date().getTime();
    firebase.storage().ref().child(fileName).putString(guess, "data_url");
  }

  await (firebase
    .firestore()
    .doc(`games/${gameCode}/notepads/${currentNotepadId}`) as DocumentReference<
    Partial<Notepad>
  >).set({ pages }, { merge: true });
}, 500);
export const submitGuess = async () => {
  const {
    client: { user },
    firebase: {
      game: { id: gameCode },
      players,
    },
  } = store.getState();
  const { currentNotepad, nextPlayer } = players[user.uid];
  const { queue: nextPlayerQueue } = players[nextPlayer];

  await (firebase
    .firestore()
    .doc(`games/${gameCode}/players/${nextPlayer}`) as DocumentReference<
    Partial<Player>
  >).set({ queue: [...nextPlayerQueue, currentNotepad] }, { merge: true });
  await (firebase
    .firestore()
    .doc(`games/${gameCode}/players/${user.uid}`) as DocumentReference<
    Partial<Player>
  >).set({ currentNotepad: "" }, { merge: true });
};

export function updateSettings(code: string, settings: Settings) {
  if (!code) return;

  client.publish({
    destination: `/app/games/${code}/update`,
    body: JSON.stringify(settings),
  });
}

export function connectToServer(uuid: string) {
  client.connectHeaders = { uuid };
  client.activate();
}

export function createGame(code: string) {
  client.publish({ destination: `/app/games/${code}/create` });
}

export async function joinGame(
  code: string,
  settings: Settings,
  rejoining: boolean = false
) {
  if (!rejoining) {
    client.publish({
      destination: `/app/games/${code}/join`,
      body: JSON.stringify(settings),
    });
  }

  // A race condition happens where if we dont wait 1ms, the subscription may happen first
  // giving an error that the game doesn't exist
  await new Promise((resolve) => setTimeout(resolve, 1));

  client.subscribe(`/topic/games/${code}`, ({ body }) => {
    const response = JSON.parse(body) as Game;
    store.dispatch(actions.handleGameUpdate(response));
  });
}
