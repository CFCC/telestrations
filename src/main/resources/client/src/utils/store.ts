import {
  configureStore,
  createAsyncThunk,
  createSlice,
  PayloadAction,
} from "@reduxjs/toolkit";
import {
  TypedUseSelectorHook,
  useSelector as useUntypedSelector,
} from "react-redux";

import * as api from "./api";
import { Game, GameState, Settings } from "./types";

interface State {
  gameState: GameState;
  openGames: string[];
  toast: {
    id: number;
    title: string;
    description: string;
    status: "info" | "warning" | "success" | "error";
  };
  settings: Settings;
  currentGame: Game;
  activeContentId: string;
}
interface ThunkApi {
  state: State;
}

export const goToLobby = createAsyncThunk<void, void, ThunkApi>(
  "goToLobby",
  (_, { getState }) => {
    const { settings } = getState();
    api.connectToServer(settings.id);
  }
);

export const createAndJoinGame = createAsyncThunk<void, string, ThunkApi>(
  "createGame",
  async (code, { getState }) => {
    const { settings } = getState();
    await api.createGame(code);
    await api.joinGame(code, settings);
  }
);

export const joinGame = createAsyncThunk<void, string, ThunkApi>(
  "joinGame",
  async (code, { getState }) => {
    const { settings } = getState();
    await api.joinGame(code, settings);
  }
);

export const rejoinGame = createAsyncThunk<void, string, ThunkApi>(
  "rejoinGame",
  async (code, { getState }) => {
    if (!code) return;

    const { settings } = getState();
    await api.joinGame(code, settings, true);
  }
);

export const saveSettings = createAsyncThunk<
  Partial<Settings>,
  Partial<Settings>,
  ThunkApi
>("saveSettings", (settings, { getState }) => {
  const {
    currentGame: { code },
    settings: oldSettings,
  } = getState();

  api.updateSettings(code, { ...oldSettings, ...settings });
  Object.entries(settings).forEach(([k, v]) => {
    if (!v) return;
    localStorage.setItem(k, v.toString());
  });

  return settings;
});

export const setGuess = createAsyncThunk<void, string, ThunkApi>(
  "setGuess",
  async (content, { getState }) => {
    const { currentGame } = getState();
    await api.setGuess(currentGame.code, content);
  }
);

export const submitGuess = createAsyncThunk<void, string, ThunkApi>(
  "submitGuess",
  async (content, { getState }) => {
    const { currentGame } = getState();
    await api.submitGuess(currentGame.code, content);
  }
);

export const startGame = createAsyncThunk<void, string, ThunkApi>(
  "startGame",
  async (gameCode) => {}
);

export const { actions, reducer } = createSlice({
  name: "app",
  initialState: {
    gameState: GameState.LOGIN,
    openGames: [],
    toast: { id: 0, title: "", description: "", status: "info" },
    currentGame: {
      active: false,
      code: "",
      players: [],
      admin: "",
      round: 0,
      adminId: "",
      isDone: false,
    },
    settings: {
      id: localStorage.getItem("id"),
      avatar: localStorage.getItem("avatar"),
      name: localStorage.getItem("name"),
      connected: true,
    },
    activeContentId: "",
  } as State,
  reducers: {
    handleRequestException: (state, action: PayloadAction<string>) => {
      state.toast = {
        id: state.toast.id + 1,
        title: "Bad Request",
        description: action.payload,
        status: "error",
      };
    },
    handleServerException: (state, action: PayloadAction<string>) => {
      state.toast = {
        id: state.toast.id + 1,
        title: "Server Error",
        description: action.payload,
        status: "error",
      };
    },
    handleSuccess: (state, action: PayloadAction<string>) => {
      state.toast = {
        id: state.toast.id + 1,
        title: "",
        description: action.payload,
        status: "success",
      };
    },
    handleGamesListMessage: (state, action: PayloadAction<string[]>) => {
      state.openGames = [...action.payload];
    },
    handleGameUpdate: (state, action: PayloadAction<Game>) => {
      state.currentGame = action.payload;
    },
    viewPlayerHistory: (state, action: PayloadAction<string>) => {
      state.gameState = GameState.PLAYER_HISTORY;
      state.activeContentId = action.payload;
    },
    viewNotepadHistory: (state, action: PayloadAction<string>) => {
      state.gameState = GameState.NOTEPAD_HISTORY;
      state.activeContentId = action.payload;
    },
    setGameState: (state, action: PayloadAction<GameState>) => {
      state.gameState = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(saveSettings.fulfilled, (state, action) => {
        state.settings = {
          ...state.settings,
          ...action.payload,
        };
      })
      .addCase(rejoinGame.fulfilled, (state, action) => {
        state.gameState = GameState.IN_GAME;
      });
  },
});

export const store = configureStore({ reducer });

export const useSelector: TypedUseSelectorHook<State> = useUntypedSelector;
