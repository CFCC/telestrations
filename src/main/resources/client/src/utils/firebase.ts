import _ from "lodash";
import { v4 as uuid } from "uuid";
import { animals, colors, uniqueNamesGenerator } from "unique-names-generator";

import { clientSlice, firebaseSlice, GameState, store } from "./store";

const {
  actions: { updatePlayers, updateNotepads, updateGame },
} = firebaseSlice;
const {
  actions: { setGameState, setUser },
} = clientSlice;

firebase.auth().onAuthStateChanged(function (user: firebase.User | null) {
  if (!user) return;

  let { displayName, uid } = user;
  if (!user.displayName) {
    displayName = _.startCase(
      uniqueNamesGenerator({
        dictionaries: [colors, animals],
        length: 2,
        separator: "-",
      })
    );
    user.updateProfile({ displayName });
  }

  store.dispatch(setUser({ uid, displayName }));
});

export const getImageURL = async (image: string): Promise<string> => {
  return (await firebase.storage().ref(image).getDownloadURL()) as string;
};
export const getAllGameCodes = async () => {
  return (await firebase.firestore().collection("games").get()).docs;
};
export const getGameCodes = (callback: (ids: string[]) => void) => {
  return firebase
    .firestore()
    .collection("games")
    .where("status", "==", "lobby")
    .onSnapshot(async function (snapshot) {
      const newGames = snapshot.docs.map((doc) => doc.id);
      callback(newGames);
    });
};
export const setGameCode = (gameCode: string, isClient: boolean = false) => {
  const game = firebase
    .firestore()
    .doc(`games/${gameCode}`) as DocumentReference<Partial<Game>>;

  game.onSnapshot(async (snapshot) => {
    const newGame = snapshot.data();
    if (!newGame) return;

    if (isClient) {
      if (newGame.status === "in progress") {
        store.dispatch(setGameState(GameState.IN_GAME));
      } else if (newGame.status === "finished") {
        store.dispatch(setGameState(GameState.FINISHED));
        const currentUser = firebase.auth().currentUser;
        if (currentUser?.isAnonymous) {
          await currentUser.delete();
        }
      }
    }

    store.dispatch(updateGame({ ...newGame, id: snapshot.id }));
  });
  game.collection("notepads").onSnapshot((snapshot) => {
    const notepads = _.mapValues(
      _.keyBy(snapshot.docs, "id"),
      (d) => d.data() as Notepad
    );
    store.dispatch(updateNotepads(notepads));
  });
  game.collection("players").onSnapshot(async (snapshot) => {
    const players = _.mapValues(
      _.keyBy(snapshot.docs, "id"),
      (d) => d.data() as Player
    );
    store.dispatch(updatePlayers(players));

    if (!isClient) {
      const {
        notepads,
        game: { status },
      } = store.getState().firebase;
      const gameIsOver =
        status === "in progress" &&
        Object.entries(players).every(
          ([pid, p]) => notepads[p.currentNotepad]?.ownerId === pid
        ) &&
        Object.values(notepads).every((n) => n.pages.length > 1);

      if (gameIsOver) await game.set({ status: "finished" }, { merge: true });
    } else {
      const { uid } = store.getState().client.user;
      const { currentNotepad, queue } = players[uid];

      if (!currentNotepad && queue?.length > 0) {
        const newQueue = [...queue];
        const newNotepad = newQueue.shift();

        await (firebase
          .firestore()
          .doc(`games/${gameCode}/players/${uid}`) as DocumentReference<
          Partial<Player>
        >).set(
          { queue: newQueue, currentNotepad: newNotepad },
          { merge: true }
        );
        store.dispatch(clientSlice.actions.setGameState(GameState.IN_GAME));
      }
    }
  });
};
export const createGame = async (gameCode: string) => {
  const {
    client: { user },
  } = store.getState();

  await (firebase
    .firestore()
    .doc(`games/${gameCode}`) as DocumentReference<Game>).set({
    created: new Date().getTime(),
    status: "lobby",
    serverId: user.uid,
  } as Game);
};
export const joinGame = async (gameCode: string) => {
  const {
    client: { user },
  } = store.getState();

  await (firebase
    .firestore()
    .doc(`games/${gameCode}/players/${user.uid}`) as DocumentReference<
    Partial<Player>
  >).set({ name: user.displayName as NonNullable<string> });
};
export const startGame = async () => {
  const {
    firebase: {
      game: { id: gameCode },
      players,
    },
  } = store.getState();
  const playerIds = Object.keys(players);

  await Promise.all(
    playerIds.map(async (playerId, i) => {
      const notepadId = uuid();
      await (firebase
        .firestore()
        .doc(
          `games/${gameCode}/notepads/${notepadId}`
        ) as DocumentReference<Notepad>).set({ ownerId: playerId, pages: [] });
      await (firebase
        .firestore()
        .doc(`games/${gameCode}/players/${playerId}`) as DocumentReference<
        Partial<Player>
      >).set(
        {
          currentNotepad: notepadId,
          nextPlayer: playerIds[(i + 1) % playerIds.length],
          queue: [],
        },
        { merge: true }
      );
    })
  );
  await (firebase.firestore().doc(`games/${gameCode}`) as DocumentReference<
    Partial<Game>
  >).set({ status: "in progress" }, { merge: true });
};
const updateGuessInFirebase = _.throttle(async (guess: string) => {
  const {
    client: { user },
    firebase: {
      game: { id: gameCode },
      players,
      notepads,
    },
  } = store.getState();
  const currentNotepadId = players[user.uid].currentNotepad;
  const currentNotepad = notepads[currentNotepadId];
  if (!currentNotepad) return;

  const pages = _.cloneDeep(currentNotepad.pages);
  const needsNewPage = _.last(pages)?.author !== user.uid;
  if (needsNewPage) {
    pages.push({
      author: user.uid,
      lastUpdated: new Date().getTime(),
      content: "",
    });
  }

  if (pages.length % 2 === 1) {
    pages[pages.length - 1].content = guess;
    pages[pages.length - 1].lastUpdated = new Date().getTime();
  } else {
    let fileName = _.last(pages)?.content;
    if (!fileName) {
      fileName = `${uuid()}.png`;
      pages[pages.length - 1].content = fileName;
    }

    pages[pages.length - 1].lastUpdated = new Date().getTime();
    firebase.storage().ref().child(fileName).putString(guess, "data_url");
  }

  await (firebase
    .firestore()
    .doc(`games/${gameCode}/notepads/${currentNotepadId}`) as DocumentReference<
    Partial<Notepad>
  >).set({ pages }, { merge: true });
}, 500);
export const setGuess = (guess: string) => updateGuessInFirebase(guess);
export const submitGuess = async () => {
  const {
    client: { user },
    firebase: {
      game: { id: gameCode },
      players,
    },
  } = store.getState();
  const { currentNotepad, nextPlayer } = players[user.uid];
  const { queue: nextPlayerQueue } = players[nextPlayer];

  await (firebase
    .firestore()
    .doc(`games/${gameCode}/players/${nextPlayer}`) as DocumentReference<
    Partial<Player>
  >).set({ queue: [...nextPlayerQueue, currentNotepad] }, { merge: true });
  await (firebase
    .firestore()
    .doc(`games/${gameCode}/players/${user.uid}`) as DocumentReference<
    Partial<Player>
  >).set({ currentNotepad: "" }, { merge: true });
};
