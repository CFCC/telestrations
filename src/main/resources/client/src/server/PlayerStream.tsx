import React, {useEffect, useState} from "react";
import {CardContent, Typography} from "@material-ui/core";
import styled from "styled-components";

import {useSelector} from "../utils/store";
import {getImageURL} from "../utils/firebase";

interface PlayerStreamProps {
    playerId: string;
}

interface PictureContainerProps {
    picture: string;
}

const PictureContainer = styled.div`
    background-image: ${(props: PictureContainerProps) => props.picture === "/question-marks.jpg" ? "url(/question-marks.jpg)" : "none"};
    background-repeat: repeat;
    width: 100%;
    min-height: ${(props: PictureContainerProps) => props.picture === "/question-marks.jpg" ? "300px" : "0"};
`;

const Picture = styled.img`
    max-width: 100%;
    min-width: 100%;
    height: auto;
`;

const Content = styled(CardContent)`
    padding-bottom: 0;
`;

export default function PlayerStream({playerId}: PlayerStreamProps) {
    const {players, notepads} = useSelector(state => state.firebase);
    const [pictureUrl, setPictureUrl] = useState("");

    const notepad = notepads[players[playerId]?.currentNotepad];
    const playerIndexInNotepad = notepad?.pages?.length;

    let picture: string, t, text;
    if (playerIndexInNotepad == null) {
        picture = "/question-marks.jpg";
        text = "Waiting for next notepad...";
    } else if (playerIndexInNotepad <= 1) {
        picture = "/question-marks.jpg";
        text = notepad.pages[playerIndexInNotepad - 1]?.content;
    } else if (playerIndexInNotepad % 2 === 0) {
        picture = notepad.pages[playerIndexInNotepad - 1]?.content;
        t = notepad.pages[playerIndexInNotepad - 1]?.lastUpdated;
        text = notepad.pages[playerIndexInNotepad - 2]?.content;
    } else {
        picture = notepad.pages[playerIndexInNotepad - 2]?.content;
        t = notepad.pages[playerIndexInNotepad - 2]?.lastUpdated;
        text = notepad.pages[playerIndexInNotepad - 1]?.content;
    }

    useEffect(() => {
        (async function() {
            const url = await getImageURL(picture);
            setPictureUrl(url);
        })();
    }, [picture, t]);

    return (
        <React.Fragment>
            <PictureContainer picture={picture}>
                {picture !== "/question-marks.jpg" && (
                    <Picture
                        src={pictureUrl}
                        alt={text}
                    />
                )}
            </PictureContainer>
            <Content>
                <Typography align="center">{text}</Typography>
            </Content>
        </React.Fragment>
    );
}
