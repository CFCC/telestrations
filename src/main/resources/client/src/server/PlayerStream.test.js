import { render } from "../utils/render";
import { WhileDrawing, WhileTyping } from "./PlayerStream.stories";
import { act } from "@testing-library/react";

it("While Drawing Renders", async () => {
  let result;
  await act(() => {
    result = render(WhileDrawing());
  });

  expect(result.asFragment()).toMatchSnapshot();
});

it("While Typing Renders", async () => {
  let result;
  await act(() => {
    result = render(WhileTyping());
  });

  expect(result.asFragment()).toMatchSnapshot();
});
