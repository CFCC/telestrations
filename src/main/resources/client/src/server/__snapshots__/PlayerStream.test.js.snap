// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`While Drawing Renders 1`] = `
<DocumentFragment>
  <div
    class="sc-bdnxRM irttLM"
  />
  <div
    class="MuiCardContent-root sc-dlnjwi kQbQio"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 MuiTypography-alignCenter"
    >
      Waiting for next notepad...
    </p>
  </div>
</DocumentFragment>
`;

exports[`While Typing Renders 1`] = `
<DocumentFragment>
  <div
    class="sc-bdnxRM irttLM"
  />
  <div
    class="MuiCardContent-root sc-dlnjwi kQbQio"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 MuiTypography-alignCenter"
    >
      Waiting for next notepad...
    </p>
  </div>
</DocumentFragment>
`;
